source/extensions/common/wasm/context.cc

```diff
@@ -1100,12 +1100,21 @@ const Network::Connection* Context::getConnection() const {
 }
 
 WasmResult Context::setProperty(std::string_view path, std::string_view value) {
+  return setPropertyHelper(CelStateKeyPrefix, path, value);
+}
+
+WasmResult Context::setEnvoyFilterState(std::string_view path, std::string_view value) {
+  return setPropertyHelper(std::string_view(), path, value);
+}
+
+WasmResult Context::setPropertyHelper(std::string_view path_prefix, std::string_view path,
+                                      std::string_view value) {
   auto* stream_info = getRequestStreamInfo();
   if (!stream_info) {
     return WasmResult::NotFound;
   }
   std::string key;
-  absl::StrAppend(&key, CelStateKeyPrefix, toAbslStringView(path));
+  absl::StrAppend(&key, path_prefix, toAbslStringView(path));
   CelState* state = stream_info->filterState()->getDataMutable<CelState>(key);
```

Comment:
tcp_proxy doesn't use CelState so this won't work if you want to support modification of tcp_proxy filter state.

Reply:
Is there a generic way to do this? Or perhaps we could expose a very specific `set_per_connection_cluster` which sets the filter state for tcp_proxy? WDYT?

Reply:
There's no generic way to do this. I think it'd help to have a registry that maps from the filter state key to the object constructor. We could probably reuse the factory registries from extensions. WDYT?

Reply:
Let me put out a PR today or tomorrow and I'll CC you on it.
/wait

Reply:
Looking forward to your PR. Meanwhile I created a specific `set_per_connection_cluster` foreign function and wrote a wasm golang plugin and it works to select upstream cluster in `tcp_proxy` (based on proxy protocol tlv headers which are emitted as dynamic metadata and I can read through wasm). However I would love to see this functionality exposed through wasm officially in envoy too.

Reply:
This [PR](https://github.com/envoyproxy/envoy/pull/29030) is what I have in mind - tcp_proxy registers a factory that Wasm then looks up based on the key name.

Reply:
Thanks @kyessenov for the PR. I'll rebase my PR based on your change sometime today/tomorrow.

Reply:
/wait
I think you should use the factory to create the filter state for tcp_proxy, from the PR.

Reply:
Sorry I got caught up in some work. Will update this change as suggested by end of this week.

Reply:
I messed up the rebase so opened a new PR here - https://github.com/envoyproxy/envoy/pull/29320